### PR TITLE
Feature/migration-nx

### PR DESCRIPTION
**Pull request by Nicoss54**
This PR add a lot of feature and improve students experience:

- Migrate Angular and Angular Matererial in last version
- Migrate all the workshop on Nx architecture
- Migrate all slides on Nx architecture
- Improve Archtecture of all workshop to add core, feature and shared module and go to atomic architecture
- Type all exercice
- Update all solution to have a solution with flux rxjs management
- add readme step by step to let students to make exercice with more simplicity
- add a new exercice with elf state management
- add new chapter on elf state management
- improve test experience with jest 
- improve gitlab ci to avoid regression
- add script to simplify mental model for students

### CODE REVIEW

**Review by tinesoft — COMMENTED** [review]
Looks good overall.

A few remarks though:

In the `package.json`, some dependencies should be moved to `devDependencies`:
* "chokidar-cli": "^1.2.2",
* "cpx": "^1.5.0",
* "parallelshell": "3.0.1",
* "sass": "^1.19.0",
* "snowpack": "^1.6.0"

Also, use `git mv` whenever possible to keep the file history (`TP/` --> `apps`)

And final note, try to make smaller PRs next time!

a PR containing 6k+ files is *very hard to review* and freezes reviewing tools (GitHub or even local a local VSCode).

File: .github/workflows/tp.yml [hunk 1]
```diff
@@ -28,8 +30,15 @@ jobs:
 
       - name: Install dependancies
```
[the remark this answers is not in the document]
Comment (Nicoss54) [reply .github/workflows/tp.yml:31]: thanks for the typo

File: .github/workflows/tp.yml [hunk 2]
```diff
@@ -9,6 +9,8 @@ jobs:
     steps:
       - name: Checkout
         uses: actions/checkout@v2
+        with:
+          fetch-depth: 0
 
       - name: Manage npm modules cache
```
[the remark this answers is not in the document]
Comment (Nicoss54) [reply .github/workflows/tp.yml:15]: ohhh tkanks for the tip

File: .github/workflows/tp.yml [hunk 3]
```diff
@@ -28,8 +30,15 @@ jobs:
 
       - name: Install dependancies
         run: npm ci
-        working-directory: TP
+
+      - name: Fetch Main Branch
```
[the remark this answers is not in the document]
Comment (Nicoss54) [reply .github/workflows/tp.yml:34]: ok so i remove this useless config

**Pull request merged**